### ISSUE TITLE
Rendered source centroid is offset from the SatNet annotation `x_center`/`y_center` by a fixed sub-pixel amount

### ISSUE THREAD

**Issue by marcodlk** [ssc-ai/satsim#35]
## Summary

The flux centroid of a rendered point source does not coincide with the position reported for it in the SatNet annotation (`x_center`, `y_center`). Across a corpus of simulated frames the offset is **fixed, sub-pixel, and axis-dependent**: in our configuration the rendered centroid sits **~+0.12 px in the row axis (axis‑0 / `y_center`)** and **<0.02 px in the column axis (axis‑1 / `x_center`)** relative to the annotation.

Because the annotation is used as astrometric/centroid ground truth by downstream tools, this injects a fixed positional bias into any sub-pixel centroid evaluation done against SatSim labels. Detection/recall is unaffected (the offset is far below a typical match radius); the problem is specifically that **the truth label and the rendered light are not co-located at the sub-pixel level.**

## Environment

- SatSim `0.23.1` (`SWCREATE = SATSIM 0.23.1` in the FITS header)
- Frame: 2048×2048, detector binning 2×2 (`XBINNING=2`, `YBINNING=2`), `EXPTIME=3.0`, `TRKMODE=rate`
- WCS: `CROTA≈-90°`, plate scale `CDELT≈0.000249444 deg/px` (≈0.898″/px)
- Objects: tasked point sources (rate-tracked → near-stationary on the FPA)

## Observed behavior

Measuring `(rendered flux centroid) − (annotation center)` for **273 stationary, isolated point sources** across 34 scenes (same sensor):

| axis | offset (px) | notes |
|------|-------------|-------|
| column / `x_center` (axis‑1) | **−0.015** | consistent with zero |
| row / `y_center` (axis‑0)    | **+0.123** | the defect |

The offset is:

- **Fixed** — independent of detector position, sub-pixel phase, SNR, and PSF FWHM.
- **Brightness-independent** — flat at +0.12–0.13 px across peak counts from ~466 to ~2590 ADU (so it is *not* an undersampling/faint-source artifact).
- **Present on stationary sources** — objects whose intra-exposure motion (`(x_start,y_start)→(x_end,y_end)`) is <0.3 px show the full offset; it does **not** correlate with the trajectory vector (Pearson r ≈ 0), so it is not an exposure-integration / trailing effect.
- **Universal** — same sign in every scene.

`x_center`/`y_center` themselves are the geometric bbox-center of the object's sampled positions (`satnet.py`, `_annotate_object`), normalized `(idx+0.5)/N` — this convention was verified correct and is not the issue; the issue is that the *rendered flux* is not centered there.

### Measurement is validated

The centroider used for the numbers above (background-subtracted 2D Gaussian fit and flux center-of-mass) recovers a **synthetic** Gaussian's known sub-pixel center to **0.000 px**, including when the synthetic is rendered on a 2×-oversampled grid and binned 2×2 (the SatSim-like path). So the +0.12 px is a property of the SatSim output, not of the measurement.

## Reproduction

1. Render a scene containing a **single static point source** at a known sub-pixel position `(row r, col c)` (rate mode, an oversample factor, 2×2 binning as above).
2. Centroid the source in the output FITS (2D Gaussian fit or windowed centroid on background-subtracted pixels).
3. Convert the annotation to a 0-based pixel: `col_truth = x_center*W − 0.5`, `row_truth = y_center*H − 0.5`.
4. Compare `centroid − truth`.

Sweeping this over a grid of sub-pixel phases, oversample factors, and binning factors will map the offset. We expect the row/col offsets to move as a function of the oversample factor and detector-bin divisibility (see root cause).

## Root-cause analysis

The rendered-flux position and the annotation position are produced by **two different coordinate paths that are not guaranteed to agree at the sub-pixel level**:

1. **Flux placement truncates the sub-pixel position.**
   `satsim/image/fpa.py`, `add_counts` (~lines 241–250):
   ```python
   r = tf.cast(r, tf.int32) + tf.cast(r_offset, tf.int32)
   c = tf.cast(c, tf.int32) + tf.cast(c_offset, tf.int32)
   ```
   Object counts are deposited at the **truncated** integer oversampled pixel — no bilinear/area interpolation and no round-to-nearest. This floors the source onto the oversampled grid (biasing toward smaller index by up to 0.5 oversampled px) and discards the sub-oversampled-pixel location. `r_offset`/`c_offset` (padding offsets) are added here and, if they differ between axes, make the residual asymmetric in row vs. column.

2. **Downsampling uses `avg_pool(..., padding='SAME')`.**
   `satsim/image/fpa.py`, `downsample` (~lines 34–47):
   ```python
   tf.nn.avg_pool(..., [1, osf, osf, 1], [1, osf, osf, 1], padding='SAME')
   ```
   With `padding='SAME'` and stride = kernel = `osf`, TensorFlow chooses padding to align the *output* shape; when the (padded, oversampled) dimension is not an exact multiple of `osf`, the implicit padding is asymmetric and shifts the binned grid origin by a sub-pixel fraction. If the row and column dimensions (after the padding/crop in `render.py`, `h_pad_os_div2` vs `w_pad_os_div2`) have different divisibility by `osf`, the induced shift **differs between axes** — matching the observed "row offset, column clean."

3. **The annotation does not pass through (1)+(2).**
   `satsim/io/satnet.py`, `_annotate_object`:
   ```python
   rr_norm = (rr + 0.5) / height
   cc_norm = (cc + 0.5) / width
   # x_center = (min(cc_norm)+max(cc_norm))/2, etc.
   ```
   `rr`/`cc` are the object's positions in **final-pixel** coordinates; they are normalized directly and never subjected to the oversampled truncation + `SAME`-pool origin shift that the *flux* undergoes. So the label tracks the requested position while the rendered light tracks the truncated+re-binned position — hence a fixed sub-pixel disagreement.

(The detector 2×2 bin compounds with the render oversample `s_osf`, which is why the magnitude is expected to depend on both.)

## Proposed solution

Make flux placement and the annotation share one sub-pixel-accurate coordinate convention:

- **Sub-pixel-accurate deposition** in `add_counts`: replace the `tf.cast(...)` truncation with area/bilinear deposition of counts across the four neighboring oversampled pixels (or, minimally, round-to-nearest). This makes the rendered centroid equal the requested float position up to the PSF, independent of oversample factor.
- **Deterministic, shift-free binning** in `downsample`: prefer an explicit reshape-and-mean — `x.reshape(H//s, s, W//s, s).mean(axis=(1,3))` (or `skimage.measure.block_reduce`) — over `avg_pool(padding='SAME')`, and ensure the oversampled+padded dimensions are exact multiples of `s` so no implicit origin shift is introduced. If `avg_pool` is retained, use `padding='VALID'` on pre-cropped, exactly-divisible dimensions and verify row/col symmetry.
- **Single source of truth for position**: have `_annotate_object` derive `x_center/y_center` from the *same* transform used to place flux (same origin, same oversample→bin mapping), so `truth == rendered centroid` by construction.
- **Regression test** (`tests/test_render.py` / `tests/test_fpa.py`): render a single static point at a grid of sub-pixel `(r,c)` phases and oversample/bin factors; assert the measured flux centroid equals the annotation `(y_center·H−0.5, x_center·W−0.5)` within a tight tolerance (e.g. 0.02 px), on **both** axes.

## Impact

- Any consumer treating SatNet `x_center`/`y_center` as sub-pixel astrometric truth inherits a fixed ~0.1–0.2 px positional bias (≈0.1–0.2″ at 0.9″/px), which corrupts centroid-accuracy benchmarking and per-axis error budgeting.
- Object detection/classification metrics are unaffected (offset ≪ match radius).
- The magnitude/axis of the offset is configuration-dependent (oversample factor, binning, padding), so it is not a constant that downstream users can simply subtract.
